dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -598,10 +598,14 @@ void DistributedBlockOutputStream::writeToShard(const Block & block, const std::
             /// Add new fields here, for example:
             /// writeVarUInt(my_new_data, extra_info);
 
-            writeVarUInt(DBMS_DISTRIBUTED_SIGNATURE_EXTRA_INFO, out);
-            writeStringBinary(extra_info.str(), out);
+            writeStringBinary(query_string, extra_info);
 
-            writeStringBinary(query_string, out);
+            const String & extra_info_str = extra_info.str();
+            CityHash_v1_0_2::uint128 checksum = CityHash_v1_0_2::CityHash128(extra_info_str.c_str(), extra_info_str.size());
+
+            writeVarUInt(DBMS_DISTRIBUTED_SIGNATURE_EXTRA_INFO_WITH_CHECKSUM, out);
+            writePODBinary(checksum, out);
+            writeStringBinary(extra_info_str, out);
```

Comment:
We can make it without changing the binary format again. Something like this:

```
            WriteBufferFromOwnString extra_info;
            writeVarUInt(ClickHouseRevision::get(), extra_info);
            context.getSettingsRef().serialize(extra_info);

            /// Added in revision DBMS_MIN_REVISION_WITH_DISTRIBUTED_SEND_CHECKSUM.
            writePODBinary(CityHash_v1_0_2::CityHash128(extra_info.stringRef().data, extra_info.stringRef().size), extra_info);
            writePODBinary(CityHash_v1_0_2::CityHash128(query_string.data(), query_string.size()), extra_info);
            
            /// Add new fields here, for example:
            /// writeVarUInt(my_new_data, extra_info);

            writeVarUInt(DBMS_DISTRIBUTED_SIGNATURE_EXTRA_INFO, out);
            writeStringBinary(extra_info.str(), out);	

            writeStringBinary(query_string, out);
```

Reply:
Indeed, but in this case there will be two checksum (but fine with me)

Reply:
And since [current revision](https://github.com/ClickHouse/ClickHouse/blob/d8fa0871b9c8b39b5c988cadb12bedbe33377340/dbms/cmake/version.cmake#L2) already changes the format ([`DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS`](https://github.com/ClickHouse/ClickHouse/blob/d8fa0871b9c8b39b5c988cadb12bedbe33377340/dbms/src/Core/Defines.h#L65)), so I guess that I should wait for revision bump, or I can bump it by myself? What is the policy for it? (since usually it is bumped by the bot)

Reply:
> And since [current revision](https://github.com/ClickHouse/ClickHouse/blob/d8fa0871b9c8b39b5c988cadb12bedbe33377340/dbms/cmake/version.cmake#L2) already changes the format ([`DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS`](https://github.com/ClickHouse/ClickHouse/blob/d8fa0871b9c8b39b5c988cadb12bedbe33377340/dbms/src/Core/Defines.h#L65)), so I guess that I should wait for revision bump, or I can bump it by myself? What is the policy for it? (since usually it is bumped by the bot)

[Revision](https://github.com/ClickHouse/ClickHouse/blob/d8fa0871b9c8b39b5c988cadb12bedbe33377340/dbms/cmake/version.cmake#L2) is bumped automatically when a new release branch (for example, 19.17, 19.18) is created.

It turns out there were no release branches created since this [extra info](https://github.com/vitlibar/ClickHouse/blob/e40c140982540c0aefb6ccbb00f2a04f2c8a3124/dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp#L594) was implemented in revision 54429 and the current revision is still 54429. So the task becomes easier: I mean it's not necessary to introduce new macros like `DBMS_MIN_REVISION_WITH_DISTRIBUTED_SEND_CHECKSUM` or `DBMS_DISTRIBUTED_SIGNATURE_EXTRA_INFO_WITH_CHECKSUM`, it would be enough to just modify reading/writing this `extra_info` without any extra checks for version.

Reply:
And it is fine that if someone will use version from master without this patch, and updates version with this patch but only on one machine from cluster things will be broken?

Reply:
Some people who use different testing versions of ClickHouse on shards will see an exception until they update all those shards. It doesn't seem to be an essential issue.

Reply:
>We can make it without changing the binary format again

And another issue with this is that query size will not be checksummed, but this is also true for the extra_info, so I guess it's fine (will be limited to 1GB anyway)